src/pool/srv_pool.c

```diff
@@ -2358,27 +2407,33 @@ pool_svc_step_up_cb(struct ds_rsvc *rsvc)
 		goto out;
 	}
 
-	if (!uuid_is_null(svc->ps_pool->sp_srv_cont_hdl)) {
-		uuid_copy(pool_hdl_uuid, svc->ps_pool->sp_srv_pool_hdl);
-		uuid_copy(cont_hdl_uuid, svc->ps_pool->sp_srv_cont_hdl);
+	if (svc->ps_global_version >= DAOS_POOL_GLOBAL_VERSION_WITH_SRV_HDLS) {
+		/* See the is_pool_from_srv comment in the "else" branch. */
+		if (uuid_is_null(svc->ps_pool->sp_srv_pool_hdl))
+			uuid_copy(svc->ps_pool->sp_srv_pool_hdl, srv_pool_hdl);
 	} else {
-		uuid_generate(pool_hdl_uuid);
-		uuid_generate(cont_hdl_uuid);
-		/* Only copy server handle to make is_from_srv() check correctly, and
-		 * container server handle will not be copied here, otherwise
-		 * ds_pool_iv_refresh_hdl will not open the server container handle.
-		 */
-		uuid_copy(svc->ps_pool->sp_srv_pool_hdl, pool_hdl_uuid);
+		if (!uuid_is_null(svc->ps_pool->sp_srv_cont_hdl)) {
+			uuid_copy(srv_pool_hdl, svc->ps_pool->sp_srv_pool_hdl);
+			uuid_copy(srv_cont_hdl, svc->ps_pool->sp_srv_cont_hdl);
+		} else {
+			uuid_generate(srv_pool_hdl);
+			uuid_generate(srv_cont_hdl);
+			/* Only copy server handle to make is_pool_from_srv() check correctly, and
+			 * container server handle will not be copied here, otherwise
+			 * ds_pool_iv_refresh_hdl will not open the server container handle.
+			 */
+			uuid_copy(svc->ps_pool->sp_srv_pool_hdl, srv_pool_hdl);
+		}
 	}
 
-	rc = ds_pool_iv_srv_hdl_update(svc->ps_pool, pool_hdl_uuid, cont_hdl_uuid);
+	rc = ds_pool_iv_srv_hdl_update(svc->ps_pool, srv_pool_hdl, srv_cont_hdl);
```

Comment:
@liw if change ds_pool_iv_srv_hdl_update()'s pool_iv_update(..., CRT_IV_SYNC_EAGER, ...) (change CRT_IV_SYNC_LAZY to CRT_IV_SYNC_EAGER), can the problem mentioned by @NiuYawei be avoided?

Reply:
@liuxuezhao, if we use an eager update, then the PS will finish stepping up only after the update completes successfully. In this time window, the PS will remain inaccessible, that is, even `dmg pool query --health-only` will not work, I think. Is that right?

Reply:
As mentioned above, a new locally synchronous but remotely asynchronous semantics would be ideal.

Reply:
the PS leader switch already with a period time, EAGER will adds one sync bcast time that seems probably will not change too much if considering the switch time?

if the case that after leader stepup, "what if an IV or whatever request reached leader and called ds_cont_hdl_rdb_lookup() before the sp_srv_cont_hdl is initialized"? Seems it possibly will cause rebuild fail?

"a new locally synchronous but remotely asynchronous semantics would be ideal."
before that available, do you think EAGER could be a temporary workaround?

Reply:
I'm not worrying about one sync bcast time; I'm thinking about the case where this bcast times out (Or hang? If we don't retry it, then no hang.). E.g., the system map is not fully in sync, or the pool map is not fully in sync.

Reply:
Has this caused rebuild issues already? If so we can discuss if there's short-term way to change the IV code to address the issues?

Reply:
> Has this caused rebuild issues already? If so we can discuss if there's short-term way to change the IV code to address the issues?

No, just by analysis that once the PS leader step up, rebuild or EC agg possibly lookup/use the handle and fail.